compiler/noirc_frontend/src/lexer/lexer.rs

```diff
@@ -376,14 +381,35 @@ impl<'a> Lexer<'a> {
         // Underscores needs to be stripped out before the literal can be converted to a `FieldElement.
         let integer_str = integer_str.replace('_', "");
 
-        let integer = match FieldElement::try_from_str(&integer_str) {
-            None => {
+        let big_integer_result = match integer_str.strip_prefix("0x") {
+            Some(integer_str) => BigInt::from_str_radix(integer_str, 16),
+            None => BigInt::from_str(&integer_str),
+        };
+
+        let integer = match big_integer_result {
+            Ok(big_integer) => {
+                if big_integer > self.max_integer {
+                    return Err(LexerErrorKind::IntegerLiteralTooLarge {
+                        span: Span::inclusive(start, end),
+                        limit: self.max_integer.to_string(),
+                    });
+                }
+                match FieldElement::try_from_str(&integer_str) {
```

Comment:
Ideally we could go from `BigInteger` to `FieldElement` without having to parse the string again... I noticed that the string "09" correctly parses as an integer (in Rust too) but `FieldElement::try_from_str` rejects it. That string literal is also in some tests that say it should be rejected. We could either error on leading zeroes, then create turn the big int into a field element (not sure how yet), or we could accept those leading zeroes (like Rust does, but maybe leading zeroes could be confused with octal numbers). Or... we could just leave it like this.

Reply:
> Ideally we could go from BigInteger to FieldElement without having to parse the string again

Have you tried using `FieldElement::from_be_bytes_reduce`?

> Or... we could just leave it like this.

I'd like to preserve the current behavior for now :+1:

Reply:
> Have you tried using FieldElement::from_be_bytes_reduce?

Nice, thanks! I just pushed a commit that uses that.

Reply:
Tests passed out of the box, so I think I understood something wrong regarding that "09" string.